Unblock UI while sharing

FSM1 opened the pull request:
Closes #1377

FSM1 committed:
WIP

actions-user committed:
lingui extract

FSM1 committed:
fix path

FSM1 committed:
add transfer toast

FSM1 committed:
update toast message

actions-user committed:
lingui extract

FSM1 committed:
fix create new shared folder

actions-user committed:
lingui extract

FSM1 committed:
fix lint

FSM1 committed:
Merge branch 'feat/unblock-ui-while-sharing-1377' of https://github.com/ChainSafe/files-ui into feat/unblock-ui-while-sharing-1377

Tbaut commented on packages/files-ui/src/Components/Modules/FileBrowsers/ShareToSharedFolderModal.tsx:
```diff
@@ -209,24 +201,17 @@ interface IShareFileProps {
   const [isUsingCurrentBucket, setIsUsingCurrentBucket] = useState(true)
-  const [keepOriginalFile, setKeepOriginalFile] = useState(true)
-  const [currentStep, setCurrentStep] = useState<Step>("1_SHARED_FOLDER_SELECTION_CREATION")
+  const [keepOriginalFile, setKeepOriginalFile] = useState(false)
```
Shouldn't we keep this true by default, to be safe?
```suggestion
  const [keepOriginalFile, setKeepOriginalFile] = useState(true)
```

Tbaut commented on packages/files-ui/src/Components/Modules/FileBrowsers/ShareToSharedFolderModal.tsx:
```diff
@@ -209,24 +201,17 @@ interface IShareFileProps {
+  const { bucket } = useFileBrowser()
   const { profile } = useUser()
-  const { getFile, error: downloadError, isDownloading } = useGetFile()
-  const [error, setError] = useState("")
```
Can we bring back the errors? I think they'd help us debug if something happens in the wild and users don't open the console.

FSM1 replied:
By the time an error occurs, the modal would be closed already, so there isn't really anywhere for it to be displayed.

Tbaut replied:
Fair enough, I guess there are very few issues that can happen with the shared folder creation, couldn't find any that will happen often.

Tbaut commented on packages/files-ui/src/Components/Modules/TransferProgressToast/TransferToast.tsx:
```diff
@@ -0,0 +1,114 @@
+    <>
+      <div className={clsx(classes.appearBox, classes.boxContainer)}
+        data-cy="upload-status-toast-message" >
+        {complete ? (
```
mixing completion and error in several ternary operations make the code pretty hard to review.
Could we change this for things like:
```
{ !!error && (< SomeErrorElement />) }
{ !complete && (< SomeProgressElement />) }
{ complete && !error && (< SomeCompleteElement />) }
```

Tbaut commented on packages/files-ui/src/Components/Modules/TransferProgressToast/index.tsx:
```diff
@@ -0,0 +1,51 @@
+  return (<div className={classes.root}>
+    {transfersInProgress.map(
+      (transfersInProgress) =>
+        (desktop || transfersInProgress.complete || transfersInProgress.error) && (
```
On mobile we don't show anything?

If we really want to keep it that way, then the `desktop` check should probably be in an early `if (!desktop) return`

Tbaut commented on packages/files-ui/src/Contexts/FilesContext.tsx:
```diff
@@ -99,6 +110,14 @@ type FilesContext = {
     writers?: UpdateSharedFolderUser[],
     readers?: UpdateSharedFolderUser[]
   ) => Promise<void>
+  transferFileBetweenBuckets: (
```
~Should we set that up for several files right away or is that too complex for a first step?~
Scratch that, I'll do it in https://github.com/ChainSafe/ui-monorepo/issues/1363

Tbaut commented on packages/files-ui/src/Contexts/FilesContext.tsx:
```diff
@@ -583,6 +621,104 @@ const FilesProvider = ({ children }: FilesContextProps) => {
+          type: "error",
+          payload: {
+            id: toastId,
+            errorMessage: "An error occured while downloading the file"
```
translation and typo
```suggestion
            errorMessage: t`An error occurred while downloading the file`
```

Tbaut commented on packages/files-ui/src/Contexts/FilesContext.tsx:
```diff
@@ -583,6 +621,104 @@ const FilesProvider = ({ children }: FilesContextProps) => {
+        type: "error",
+        payload: {
+          id: toastId,
+          errorMessage: `An error occured: ${error}`
```
translation and typo
```suggestion
          errorMessage: t`An error occurred: ${error}`
```

Tbaut commented on packages/files-ui/src/Components/Modules/FileBrowsers/hooks/useCreateOrEditSharedFolder.tsx:
```diff
@@ -25,8 +25,7 @@ export const useCreateOrEditSharedFolder = () => {
         return bucket
-      })
-      .catch((error) => {
+      }).catch((error) => {
```
nit for consistency, I think it was fine as it was.
```suggestion
      })
      .catch((error) => {
```

Tbaut commented on packages/files-ui/src/Components/Modules/TransferProgressToast/TransferToast.tsx:
```diff
@@ -0,0 +1,114 @@
+}
+
+const TransferToast: React.FC<ITransferToast> = (props) => {
+  const { transferInProgress } = props
```
nit
```suggestion
const TransferToast: React.FC<ITransferToast> = ({ transferInProgress }) => {
```

Tbaut commented on packages/files-ui/src/Components/Modules/TransferProgressToast/TransferToast.tsx:
```diff
@@ -0,0 +1,114 @@
+  return (
+    <>
+      <div className={clsx(classes.appearBox, classes.boxContainer)}
+        data-cy="upload-status-toast-message" >
```
```suggestion
      <div
        className={clsx(classes.appearBox, classes.boxContainer)}
        data-cy="upload-status-toast-message"
      >
```

Tbaut commented on packages/files-ui/src/Components/Modules/TransferProgressToast/index.tsx:
```diff
@@ -0,0 +1,51 @@
+  ITheme,
+  makeStyles,
+  useThemeSwitcher
+} from "@chainsafe/common-theme"
```
nit
```suggestion
import { createStyles, ITheme, makeStyles, useThemeSwitcher } from "@chainsafe/common-theme"
```

Tbaut commented on packages/files-ui/src/Contexts/FilesReducers.tsx:
```diff
@@ -117,3 +117,73 @@ export function downloadsInProgressReducer(
+    if (progressIndex > -1) {
+      transfersInProgress[progressIndex].error = true
+      transfersInProgress[progressIndex].errorMessage =
+          action.payload.errorMessage
```
```suggestion
      transfersInProgress[progressIndex].errorMessage = action.payload.errorMessage
```

Tbaut commented on packages/files-ui/src/Components/Modules/TransferProgressToast/TransferToast.tsx:
```diff
@@ -0,0 +1,114 @@
+              className={classes.marginBottom}
+            >
+              <Trans>
+                Transferring your file ({operation})
```
The operation is a non translated string, so it will always appear in english. I guess it's not a big deal for now

Tbaut commented on packages/files-ui/src/Contexts/FilesContext.tsx:
```diff
@@ -583,6 +621,104 @@ const FilesProvider = ({ children }: FilesContextProps) => {
+    sourceFile: FileSystemItem,
+    path: string,
+    destinationBucket: BucketKeyPermission,
+    keepOriginal = false
```
We should default to something safe
```suggestion
    keepOriginal = true
```

Tbaut reviewed (commented):
Working very well :tada: 

Playing with it made me feel that the UX is a bit too simple for now, we should create issues to:
- tell what files are being moved, where
- allow to cancel the download/upload

edit: done in #1447

asnaith commented:
Working really well on desktop with the toast notifications.

In my opinion it does seem like a downgraded experience on **mobile** though by not showing anything at all. If the user went to check a shared folder whilst in process of sharing a large file to it then might lead to confusion if it's not there yet, with the blocked UI everyone know's for certain something is still in progress.

Tbaut committed:
Merge branch 'dev' into feat/unblock-ui-while-sharing-1377

FSM1 committed:
Update packages/files-ui/src/Components/Modules/FileBrowsers/ShareToSharedFolderModal.tsx

Co-authored-by: Thibaut Sardan <33178835+Tbaut@users.noreply.github.com>

FSM1 committed:
Update packages/files-ui/src/Components/Modules/FileBrowsers/hooks/useCreateOrEditSharedFolder.tsx

Co-authored-by: Thibaut Sardan <33178835+Tbaut@users.noreply.github.com>

FSM1 committed:
Update packages/files-ui/src/Contexts/FilesContext.tsx

Co-authored-by: Thibaut Sardan <33178835+Tbaut@users.noreply.github.com>

FSM1 committed:
Update packages/files-ui/src/Components/Modules/TransferProgressToast/index.tsx

Co-authored-by: Thibaut Sardan <33178835+Tbaut@users.noreply.github.com>

FSM1 committed:
Update packages/files-ui/src/Components/Modules/TransferProgressToast/TransferToast.tsx

Co-authored-by: Thibaut Sardan <33178835+Tbaut@users.noreply.github.com>

FSM1 committed:
update transfer toast

FSM1 committed:
Update packages/files-ui/src/Contexts/FilesReducers.tsx

Co-authored-by: Thibaut Sardan <33178835+Tbaut@users.noreply.github.com>

FSM1 committed:
enable transfer toast for mobile

FSM1 commented:
> Working really well on desktop with the toast notifications.
> 
> In my opinion it does seem like a downgraded experience on **mobile** though by not showing anything at all. If the user went to check a shared folder whilst in process of sharing a large file to it then might lead to confusion if it's not there yet, with the blocked UI everyone know's for certain something is still in progress.

I have enabled the transfer toast on mobile as well now.

FSM1 committed:
Merge branch 'dev' into feat/unblock-ui-while-sharing-1377

Tbaut committed:
Apply suggestions from code review

tanmoyAtb reviewed (changes requested):
This is working great!

On UX level: I got this view of error when I tried to share file to shared folder that had a file with the same name. Just need to work on showing the error
<img width="479" alt="Screenshot 2021-08-17 at 10 02 59 PM" src="https://user-images.githubusercontent.com/10713353/129761103-6a5fe034-2203-4ee5-9204-3a33ff98a924.png">

Also, The word `Transferring` caught me off guard, I thought this was moving the file.
Can we just use `Sharing your file`.
(I really like the downloading -> encrypt and uploading progress texts!)

On a separate note:
If you navigate to the `Shared` page, while you're sharing a file, the progresses go away. 
Can we move the progress states higher to have them in all browsers. This can be a separate ticket

Tbaut committed:
Merge branch 'dev' into feat/unblock-ui-while-sharing-1377

actions-user committed:
lingui extract

asnaith commented:
> I have enabled the transfer toast on mobile as well now.

On the mobile view I'm seeing the toast when the transfer is complete (or if it fails) now but nothing whilst it is in progress

FSM1 committed:
Merge branch 'feat/unblock-ui-while-sharing-1377' of https://github.com/ChainSafe/files-ui into feat/unblock-ui-while-sharing-1377

FSM1 commented:
> > I have enabled the transfer toast on mobile as well now.
> 
> On the mobile view I'm seeing the toast when the transfer is complete (or if it fails) now but nothing whilst it is in progress

Hmm, seems to be working...
![image](https://user-images.githubusercontent.com/12774278/129803071-ac45b9f9-a4aa-4b7e-a607-ba29e0e6a235.png)

asnaith commented:
> Hmm, seems to be working...

👍 Yep, all working great now, seeing the modal whilst it's in progress now. Perhaps it was because I was testing it before a89a136.

FSM1 committed:
fix error message and rendering

actions-user committed:
lingui extract

FSM1 committed:
fix lint

FSM1 committed:
ensure toasts remain while navigating

FSM1 committed:
fix lint

FSM1 commented:
> This is working great!
> 
> On UX level: I got this view of error when I tried to share file to shared folder that had a file with the same name. Just need to work on showing the error
> <img alt="Screenshot 2021-08-17 at 10 02 59 PM" width="479" src="https://user-images.githubusercontent.com/10713353/129761103-6a5fe034-2203-4ee5-9204-3a33ff98a924.png">
> 
> Also, The word `Transferring` caught me off guard, I thought this was moving the file.
> Can we just use `Sharing your file`.
> (I really like the downloading -> encrypt and uploading progress texts!)
> 
> On a separate note:
> If you navigate to the `Shared` page, while you're sharing a file, the progresses go away.
> Can we move the progress states higher to have them in all browsers. This can be a separate ticket

I fixed all the error messages, updated the copy on the toast, and also moved the toast rendering up to the FilesContext level.

FSM1 committed:
Merge branch 'dev' into feat/unblock-ui-while-sharing-1377

FSM1 committed:
Merge branch 'dev' into feat/unblock-ui-while-sharing-1377

tanmoyAtb reviewed (approved):
Amazing work !

Tbaut committed:
Merge branch 'dev' into feat/unblock-ui-while-sharing-1377
